mitodl/mit-open
# Channel types

### What are the relevant tickets?





### Description (What does it do?)
Closes #685 

- Adds a few new fields to the `FieldChannel` model, including `channel_type` with 4 possible values
- Adds 4 new models to hold specific details about each channel type, each with a one-to-one relation to the FieldChannel model.  Similar to LearningResource and Course, Program etc.
- Makes the `name` and `channel_type` fields unique together (formerly `name` was unique on its own)
- Changes the existing channel api urls, views, permissions to use the id instead of the name as the channel lookup field
- Adds a new read-only view allowing a channel to be retrieved by channel_type and name instead of id.
- Updates the front-end code to work with the new/modified endpoints and urls.
- Adds a radio selection for channel_type to the channel edit page.


### Screenshots (if appropriate):

![Screenshot 2024-04-04 at 9 44 55 AM](https://github.com/mitodl/mit-open/assets/187676/53b7f403-a933-4079-b192-d17578a08a7c)


### How can this be tested?
There is no frontend channel creation page yet, so use the API at http://localhost:8063/api/v0/channels/ and post something like the following:

```json
{
    "name": "biology",
    "title": "Department of Biology",
    "public_description": "The MIT Dept of Biology",
    "channel_type": "department",
    "search_filter": "department=Biology",
    "configuration": {"foo": "bar"},
    "department_detail": {
        "department": "7"
    }
}
```

The channel should be created successfully.  
You should be able to see the details at both these URL's:
  - `http://localhost:8063/api/v0/channels/<channel_id>/` (intended for CRUD)
  - `http://localhost:8063/api/v0/channels/type/department/physics/` (read-only, for front end lookup by channel type and name)

You should be able to access the front end page for this channel at http://localhost:8063/c/department/physics/

Logged in as a superuser, click the gear icon at top right and select "Field Settings".  You should get to an edit page that has radio buttons for selecting the channel type.  Switch to another type (plus make additional changes if you want) and save.  You should be redirected back to the view page, and the url should now be different, including the channel type you just picked.

Click the gear icon at top right again and select "Manage Widgets".  You should be able to add/edit/delete widgets successfully.


### Additional Context
There are lots of files, variables, and some text in the code named `field*` that should probably be changed to `channel` for consistency, but I'm leaving that for another PR since this one is already large.

Did not add the  new `search_filter` field to the front-end edit form (that will be included in another PR).  

No edit form fields for specifying channel-specific details yet (like which topic/department/offeror if any to associate with) - will save that for a subsequent PR.  

Also did not add a front-end form field for `configuration` because what exactly is going to go in there is still up in the air, will likely depend on how we decide to add custom content components and their placement to these pages.
